src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -193,6 +198,51 @@ void handle_call(
                        address_of_exprt{function_pointer_target_symbol_expr}}));
   }
 }
+
+void get_by_name_restriction(
+  const goto_functiont &goto_function,
+  const function_pointer_restrictionst::restrictionst &by_name_restrictions,
+  function_pointer_restrictionst::restrictionst &restrictions,
+  const goto_programt::const_targett &location)
+{
+  PRECONDITION(location->is_function_call());
+
+  const exprt &function = location->get_function_call().function();
+
+  if(!can_cast_expr<dereference_exprt>(function))
+    return;
+
+  auto const &function_pointer_call_site = to_symbol_expr(
+    to_dereference_expr(function)
+      .pointer());
+
+  for(auto it = std::prev(location);
+      it != goto_function.body.instructions.end();
+      ++it)
+  {
+    if(!it->is_assign())
+      continue;
+
+    if(it->get_assign().lhs() != function_pointer_call_site)
+      continue;
+
+    if(!can_cast_expr<symbol_exprt>(it->get_assign().rhs()))
+      continue;
+
+    auto const &rhs = to_symbol_expr(it->get_assign().rhs());
+    auto const restriction =
+      by_name_restrictions.find(rhs.get_identifier());
+
+    if(
+      restriction != by_name_restrictions.end() &&
+      restriction->first == rhs.get_identifier())
+    {
+      restrictions.emplace(
+        function_pointer_call_site.get_identifier(),
+        restriction->second);
+    }
```

Comment:
My understanding is if we have

```
(*fp)()
```

Then we look for lines that look like
```
fp = <something>
```

On the proceeding line, or any line after.

I still think we should be searching backwards, but this still seems off. What if the call is on the actual function we have a named restriction for? I realize from docs that this is a simple case data flow analysis, but is there not already some code that does a data flow analysis that provides this?

Reply:
The loop was actually unnecessary as the assignment will always be on the previous line. I've removed the loop.